[RNMobile] Editor Onboarding: "The Basics" help section - Details Part 1

AmandaRiu opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->
**Gutenberg-Mobile PR**: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3777

## Description
<!-- Please describe what you have changed or added -->
This PR is the second in a series of PRs to add the Editor Onboarding basic help section defined in this ticket: https://github.com/WordPress/gutenberg/issues/30626 and covers the following:
- Adds the "details" pages for each of the basic help topics including localized text and static images (additional notes below)
Note that this PR only covers a baseline for this area of the feature and will continue to be iterated on by another developer. I'm opening this PR now to ensure the work done so far so it can neatly be handed off to the next developer as I'll be moving projects. 

**Special notes:**
- The images provided are not localized and will need to be replaced with a mechanism for creating localized images.
- Since the images will eventually be replaced before this feature rolls out to production, I have no included different sizes for different screen resolutions. 
- Does not include any considerations for accessibility
- It would make sense to make the detail page full screen in the future if needed.
- For some reason, the images would not appear if the Demo app was built directly in the gutenberg repo, but would appear if built from the gutenberg-mobile repo. @dcalhoun wrote a patch to fix this issue and I have included it in dce0398dbd4869f84d4b8701da08ac568d411e9d. 
- I attempted to use `BottomSheet.SubSheet` for the details pages, but for some reason the images would flicker uncontrollably while scrolling on iOS. I attempted to use `useMemo` to prevent this, but couldn't get it to work. For that reason I went back to my original custom implementation of `BottomSheet.NavigationScreen` and use `useMemo` in `view-section.native.js` when serving up the image assets to avoid rerendering images while scrolling. This works nicely. Below are recordings of the iOS differences. 

The "before" is using `BottomSheet.SubSheet`:

https://user-images.githubusercontent.com/5810477/127734916-90177e66-8c45-4d48-9039-b315245951ba.mp4

After is using the custom `BottomSheet.NavigationSheet`:

https://user-images.githubusercontent.com/5810477/127734945-d82e55bc-136d-4937-8ac9-666bc9ec729e.mp4

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
1. Build and run the app on iOS and Android
2. Open the menu (shake on Android) and tap on "Help". The Help "Basics" list of topics will be shown.
3. Tap on each top and verify the appropriate bottom sheet is displayed. 

## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/5810477/127735228-a36cf2ec-9ab7-420d-865b-a7dcb2b31e0d.mp4



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Commit:
Add help menu option to iOS demo app

Commit:
Add basic help topics bottom sheet

Commit:
Add help option to Android demo app

Commit:
Add new help topics icons

Commit:
Add help topics to main help bottom sheet

Commit:
Remove unused imports

Commit:
Optimized new SVG images using https://jakearchibald.github.io/svgomg/

Commit:
Merge branch 'trunk' into rnmobile/editor-onboarding-basic-help-2

Commit:
Add navigation to help detail subsheet

Commit:
Add react-native and react-navigation to package.json per lint

Commit:
Fix lint errors

Commit:
Add images for help detail sections

Commit:
Add intro to blocks page

Commit:
Add remaining help detail pages

Commit:
Link help details to topics

Commit:
WIP

Commit:
Merge branch 'trunk' into rnmobile/editor-onboarding-basic-help-4

Commit:
Remove unnecessary dependency

Commit:
Cleanup code, remove debugging

Commit:
Patch to correct issues with asset loading at runtime

Commit:
Fix lint errors by removing unused imports

dcalhoun commented on packages/editor/src/components/editor-help/add-blocks.native.js:
```diff
@@ -0,0 +1,35 @@
+ * Internal dependencies
+ */
+import styles from './style.scss';
+import { HelpDetailBodyText, HelpDetailImage } from './view-sections.native';
```
I believe we shouldn't need to append the `.native` suffix ourself and generally rely upon Metro to manage this. Also, this extra suffix occurs a few more times in this PR, but I did not call them out individually to avoid creating unnecessary noise. We should likely update those locations as well. 
```suggestion
import { HelpDetailBodyText, HelpDetailImage } from './view-sections';
```

AmandaRiu replied:
Ah good point. Updated in b2b136317383f3b251ce0d39977bdad791306bbe

dcalhoun commented on packages/editor/src/components/editor-help/help-detail-navigation-screen.native.js:
```diff
@@ -1,21 +1,21 @@
 import styles from './style.scss';
 
-const HelpDetailNavigationScreen = ( { name } ) => {
+const HelpDetailNavigationScreen = ( { name, view } ) => {
```
I wondered if there might be a more descriptive name than `view` for this prop. Possibly `content` or `helpContent`? WDYT? 

Happy to leave this as-is if you disagree. 🙇🏻

AmandaRiu replied:
I like it! Updated in af2828549537bf99702695431ab8c8435788952b

dcalhoun commented on packages/editor/src/components/editor-help/view-sections.native.js:
```diff
@@ -0,0 +1,29 @@
+};
+
+export const HelpDetailImage = ( { source } ) => {
+	return useMemo(
+		() => <Image source={ source } style={ styles.helpDetailImage } />,
+		[ source ]
+	);
```
Is the use of `useMemo` an attempt to avoid the `BottomSheet.SubSheet` flickering mentioned in the PR description? Should we retain this memoization now that `BottomSheet.SubSheet` is no longer used?

AmandaRiu replied:
That's a great question. If I remove this code, then the flickering does not happen but I left it seemed like a good optimization since scrolling was causing an almost infinite reload of those image. From a performance standpoint however, I do not see any noticeable difference when interacting with the bottom sheet so I've removed it. It can be added back in the future if deemed necessary. Thanks David! Updated in b68013cd2d5c44ae550a649d54ecc2a47250e38f

dcalhoun commented on patches/react-native+0.64.0.patch:
```diff
@@ -23,3 +23,27 @@ index 52cbda4..3e3b71e 100644
++++ b/node_modules/react-native/Libraries/Image/AssetSourceResolver.js
+@@ -92,9 +92,18 @@ class AssetSourceResolver {
+    */
+   assetServerURL(): ResolvedAssetSource {
```
I believe this patch will only repair the Metro server used for local development. From my testing, the static image assets are missing when the bundle is integrated into WPAndroid and WPiOS. So, we will need to address the bundle assets in a separate PR in the future. 

That said, I believe I have found an alternative Metro server fix that is less invasive. It relies upon changes to the local `metro.config.js` file instead of patching third-party code. I plan to open a PR against this branch for your consideration.

dcalhoun replied:
https://github.com/WordPress/gutenberg/pull/33829 is the follow up PR for consideration.

AmandaRiu replied:
That's great news! I'll take a look at your PR and hold of merging this one until your PR is merged into this branch. 👍

dcalhoun reviewed (approved):
This looks good, @AmandaRiu! Thank you for putting this work together. I verified the feature works as expected on an iPhone SE and Samsung Galaxy S20. 

I agree that a fullscreen bottom sheet likely makes a lot of sense. Currently, the help section title and back button scrolls out of view when viewing longer content. Losing visibility of the back button in particular feels odd and disorienting. Hopefully we can improve this aspect in future work. 

<details><summary>Scrolling Hides Back Button</summary>

<image width="300" alt="Scrolling hides back button on iOS Help Section bottom sheet" src="https://user-images.githubusercontent.com/438664/127921579-0e92b0a1-b4b2-4579-ac72-e2e11e3a8963.PNG" />

</details>

I left a few comments that are not blockers. We could always address them in future PRs if desired.

Commit:
Remove extreneous filename extension from import

Commit:
Rename view property for clarity

Commit:
Remove memoization

dcalhoun committed:
Fix Metro local static image paths (#33829)

Loading local static assets from a different package within a monorepo
appears broken in Metro for Android. The below fixes paths to packages
within this project to include the necessary `/assets/..` that Metro's
server expects to traverse to the correct directory.

- https://git.io/JBV4e
- https://git.io/JBFon
